### PR TITLE
Validate full file path in the static file handler

### PR DESCRIPTION
**Pull request by magv**
Hopefully prevent directory traversal attacks; as discussed in issue #380.

### CODE REVIEW

**Comment by asvetlov** [comment]
The code looks good on first glance but I cannot accept it without tests.

Please push test code after functional tests for static file handling (https://github.com/KeepSafe/aiohttp/blob/master/tests/test_web_functional.py#L263-347)

**Comment by magv** [comment]
It turns out that requiring tests is a great policy, because my first patch had an embarassing error in it: `filename` was used instead of `filepath` in one critical place, thus making `StaticRoute` forbid _all_ filenames, except for those that are absolute paths...

Interestingly enough, the current static file tests would not have caught that error, because they all use absolute paths in the request URLs -- for example, on [this line](/KeepSafe/aiohttp/blob/v0.16.2/tests/test_web_functional.py#L290) `filename` is set to the absolute path, and then [over here](/KeepSafe/aiohttp/blob/v0.16.2/tests/test_web_functional.py#L267-272) `url` is set to `"/static/"` plus that absolute path.

I think that was not intended, so I changed those test and made them use short paths in the URLs (the tests still pass).

File: aiohttp/web_urldispatcher.py [hunk 1]
```diff
@@ -149,7 +149,7 @@ def __init__(self, name, prefix, directory, *,
             'GET', self.handle, name, expect_handler=expect_handler)
         self._prefix = prefix
         self._prefix_len = len(self._prefix)
-        self._directory = directory
+        self._directory = os.path.abspath(directory) + os.sep
```
Comment (asvetlov) [inline aiohttp/web_urldispatcher.py:152]: Please add check for directory like https://github.com/KeepSafe/aiohttp/blob/master/aiohttp/web_urldispatcher.py#L397
I guess to change exception from AssertionError to ValueError.

**Comment by asvetlov** [comment]
> I think that was not intended, so I changed those test and made them use short paths in the URLs (the tests still pass).

Agree.

**Comment by magv** [comment]
Like this?

**Comment by asvetlov** [comment]
Perfect! But you forgot to add unttest for that...

**Comment by magv** [comment]
I'm not really sure where to put it though. Is this OK?

**Comment by asvetlov** [comment]
Yes

**Comment by asvetlov** [comment]
Thanks!

**Pull request merged**